### ISSUE TITLE
Deny Holiday Booking

### ISSUE THREAD

**Issue by dyerry79** [meeting-room-booking-system/mrbs-code#3771]
Hi! I see that there's a portion to prevent holiday booking. Here's the config:

![image](https://github.com/user-attachments/assets/810353ea-7952-41ca-ae87-3097facfcb8b)

However, I noticed that booking on a holiday is still accepted. See image for reference.

![image](https://github.com/user-attachments/assets/663200c3-f753-4f06-93fb-7baa5a2bcc82)

It would be better perhaps if the declaration of holidays can be made in GUI. If needed, what are the files that I should look into?

**Comment by dyerry79** [comment]
my bad.... i got it working now. it should have been

`$holidays[2024] = array('2024-11-01..2024-11-02', '2024-12-25','2024-12-30');`

just the other concern now if there's a possibility to create holidays in the calendar itself. What are the files that we need to look at to begin.

**Comment by campbell-m** [comment]
> However, I noticed that this works for non-admin users. Admin user can still book even if the selected day is declared holiday in the config file.

Yes, that's correct and is the same for all policies.  Admins can do what they like and are assumed to know what they are doing.

> Just the other concern now if there's a possibility to create holidays in the calendar itself. What are the files that we need to look at to begin.

I'm afraid that would be a lot of work.

**Comment by dyerry79** [comment]
I know, right. :). If you could point to the right files only. This could be a good feature if you could incorporate it later.

**Comment by campbell-m** [comment]
Ideally most of the configuration settings should be settable via a GUI.  But this would involve creating a new database table, a new menu structure and a new set of forms for setting the variables.  You could make a quick and dirty change that would just change the holiday settings, but I leave that to you.

Alternatively you could just book out holidays in the calendar by creating bookings called "Holiday" with type "Holiday".

**Comment by dyerry79** [comment]
> Ideally most of the configuration settings should be settable via a GUI. But this would involve creating a new database table, a new menu structure and a new set of forms for setting the variables. You could make a quick and dirty change that would just change the holiday settings, but I leave that to you.

I agree, but I am okay with the other declarations having to be set in the config file except for the holidays. Staff will be assigned to the system, so I need holidays to be set via GUI. This is what I got so far and I can no longer move forward :(.

1. Arraged some items under setup for admin user only
![image](https://github.com/user-attachments/assets/cd937859-3e35-40f6-9258-43561c4bce15)

2. Managed to create the for the holidays patterned with admin.php but I really don't know if I did the right thing. Anyway, here's the screenshot
![image](https://github.com/user-attachments/assets/06b6a871-5471-486d-aca5-b331d294d952)

3. In the Datetime.php, I see the isHoliday function so what I did is 

`  public function isHoliday() : bool
  {
    if ($this->isHolidayConfig())
    {
      return true;
    }
	global $db;
	
	$iso_date = $this->format('Y-m-d');
	
	$query = "SELECT COUNT(*) FROM mrbs_holidays WHERE holiday_date = '2024-11-01'";
	$count = $db->query1($query, array($iso_date));
	
	return $count > 0;

    // This method can be extended to check other sources, eg a .ics
    // file or a database table.
    //return false;
  }`
  but i got an error

Uncaught exception 'Error' in C:\laragon\www\book\lib\MRBS\DateTime.php at line 397
Call to a member function query1() on null
#0 C:\laragon\www\book\functions_table.inc(12): MRBS\DateTime->isHoliday()
#1 C:\laragon\www\book\functions_table.inc(1691): MRBS\get_date_classes(Object(MRBS\DateTime))
#2 C:\laragon\www\book\functions_table.inc(1611): MRBS\month_room_table_innerhtml('month', 0, 2024, 11, 7, 1, 1)
#3 C:\laragon\www\book\index.php(459): MRBS\month_table_innerhtml('month', 0, 2024, 11, 7, 1, 1)
#4 {main}

Any Ideas?

> Alternatively you could just book out holidays in the calendar by creating bookings called "Holiday" with type "Holiday".
this is also an option but the above approach is more favored as of the moment

**Comment by campbell-m** [comment]
You are calling `query1()` on `$db` instead of `db()`.

**Comment by dyerry79** [comment]
my bad. thanks campbell for the prompt response. holidays can now be both declared in the config and via GUI
